### PR TITLE
Land-river two-way coupling

### PR DESCRIPTION
**Pull request by donghuix**
This development implements the land-river two-way hydrological coupling through the infiltration on 
the floodplain during the flooding period. This feature can be turned on by setting `lnd_rof_two_way=.true.`
in  **user_nl_cpl** , and `inundflag=.true.` in **user_nl_mosart**.

[BFB]

### CODE REVIEW

**Comment by rljacob** [comment]
A test that turns on the feature should be part of this PR.

**Comment by donghuix** [comment]
> A test that turns on the feature should be part of this PR.


Hi @rljacob, I just added a test with land river two-way coupling turned on.

File: components/elm/bld/namelist_files/namelist_definition.xml [hunk 1]
```diff
@@ -1979,4 +1979,13 @@ Toggle to turn on the soil erosion model.
 Toggle to turn on the soil erosion and OM pools feedback.
 </entry>
 
+<!-- ========================================================================================  -->
+<!-- Namelist options for land river coulping                                                  -->
+<!-- ========================================================================================  -->
+
+<entry id="use_lnd_rof_two_way" type="logical" category="clm_physics"
```
Comment (bishtgautam) [inline components/elm/bld/namelist_files/namelist_definition.xml:1986]: `category` should be `elm_physics`

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 2]
```diff
@@ -387,14 +401,27 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
...
 
+             !0. partition grid-level floodplain inundation volume and fraction to each column
+             if (use_lnd_rof_two_way) then
+                if (mod(get_nstep()-1,6) == 1 .or. get_nstep() <= 1) then                 
+                  h2orof(c)      = atm2lnd_vars%h2orof_grc(g) * wtgcell(c) 
+                  frac_h2orof(c) = atm2lnd_vars%frac_h2orof_grc(g) * wtgcell(c)
+                endif
+                ! TODO: add inundfrac from ocean 
+                if ( frac_h2orof(c) > 1 - fsno - frac_h2osfc(c) ) then
+                  !h2orof(c) = h2orof(c) * frac_h2orof(c) / ( 1 - frac_sno(c) - frac_h2osfc(c) )
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SoilHydrologyMod.F90:412]: remove the commented out line

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 3]
```diff
@@ -492,7 +524,31 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
              !7. remove drainage from h2osfc and add to qflx_infl
              h2osfc(c) = h2osfc(c) - qflx_h2osfc_drain(c) * dtime
              qflx_infl(c) = qflx_infl(c) + qflx_h2osfc_drain(c)
-             qflx_gross_infl_soil(c) = qflx_gross_infl_soil(c) + qflx_h2osfc_drain(c)
+             
+             !8. add drainage from river inundation to qflx_infl (land river two way coupling)
+             if (use_lnd_rof_two_way) then
+
+               h2osoi_left_vol1 = max(max(0._r8,(pondmx+watsat(c,1)*dz(c,1)*1.e3_r8-h2osoi_ice(c,1)-watmin)) - &
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SoilHydrologyMod.F90:531]: can you please simplify the code so there aren't so many nested `max` as it is hard to follow the code?

**Review by bishtgautam — COMMENTED** [review]
@donghuix I have minor comments for you to address.

**Comment by donghuix** [comment]
This new feature is turned off by default in our development.

Land-river two-way coupling represents a realistic physical process at the river land interface (floodplain) proposed by ICoM to understand the impacts of inundation on water and biogeochemical cycles. 

We documented details of the development and evaluation of this new feature in a manuscript, which is under review at JAMES. The preprint of the manuscript can be found at: [https://www.essoar.org/doi/abs/10.1002/essoar.10507802.1].

We also run the e3sm_land_developer test suites, the new development passed all the test except reporting namelist error. The namelist error is because the `use_lnd_rof_two_way=.false.` appears in **mosart_in**.

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 (hunk 3 above)
Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyMod.F90]: I added a if statement to remove one max().

**Comment by bishtgautam** [comment]
@peterdschwartz Could you please also review this PR?

File: components/elm/bld/ELMBuildNamelist.pm [hunk 4]
```diff
@@ -2321,6 +2321,7 @@ sub setup_logic_demand {
   $settings{'use_snicar_ad'}       = $nl_flags->{'use_snicar_ad'};
   $settings{'use_century_decomp'}  = $nl_flags->{'use_century_decomp'};
   $settings{'use_crop'}            = $nl_flags->{'use_crop'};
+  $settings{'use_lnd_rof_two_way'} = $nl_flags->{'use_lnd_rof_two_way'};
```
Comment (tanzeli1982) [inline components/elm/bld/ELMBuildNamelist.pm:2324]: Not sure whether it is needed.

Comment (donghuix) [reply components/elm/bld/ELMBuildNamelist.pm:2324]: I followed other feature flag to add the two-way coupling control flag here. I am not sure if this is necessary, @bishtgautam, do you know if we can delete it?

**Comment by bishtgautam** [comment]
@peterdschwartz, Checking in to see if you will have time to review this PR? Thanks.

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 5]
```diff
@@ -387,14 +401,26 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
...
              endif
 
+             !0. partition grid-level floodplain inundation volume and fraction to each column
+             if (use_lnd_rof_two_way) then
+                if (mod(get_nstep()-1,6) == 1 .or. get_nstep() <= 1) then                 
+                  h2orof(c)      = atm2lnd_vars%h2orof_grc(g) * wtgcell(c) 
+                  frac_h2orof(c) = atm2lnd_vars%frac_h2orof_grc(g) * wtgcell(c)
+                endif
+                ! TODO: add inundfrac from ocean 
+                if ( frac_h2orof(c) > 1 - fsno - frac_h2osfc(c) ) then
```
Comment (tanzeli1982) [inline components/elm/src/biogeophys/SoilHydrologyMod.F90:411]: How about frac_h2orof(c) = min(frac_h2orof(c), 1 - fsno - frac_h2osfc(c))

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 6]
```diff
@@ -387,14 +401,26 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
                 qflx_evap(c)=qflx_ev_soil(c)
              endif
 
+             !0. partition grid-level floodplain inundation volume and fraction to each column
+             if (use_lnd_rof_two_way) then
+                if (mod(get_nstep()-1,6) == 1 .or. get_nstep() <= 1) then
```
Comment (tanzeli1982) [inline components/elm/src/biogeophys/SoilHydrologyMod.F90:406]: What is the result if the user chooses different coupling time steps?

File: components/elm/src/biogeophys/SoilHydrologyType.F90 [hunk 7]
```diff
@@ -569,7 +569,12 @@ subroutine InitCold(this, bounds)
          endif
 
          ! set decay factor
-         this%hkdepth_col(c) = 1._r8/2.5_r8
```
Comment (tanzeli1982) [inline components/elm/src/biogeophys/SoilHydrologyType.F90:572]: Why changes it?

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 8]
```diff
@@ -308,22 +315,27 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
...
+          t_soisno             =>    col_es%t_soisno             , & ! Input:  [real(r8) (:,:) ]  soil temperature (Kelvin)
+
+          frac_h2osfc          =>    col_ws%frac_h2osfc          , & ! Input:  [real(r8) (:)   ]  fraction of ground covered by surface water (0 to 1)
+          frac_sno             =>    col_ws%frac_sno_eff         , & ! Input:  [real(r8) (:)   ]  fraction of ground covered by snow (0 to 1)
+          h2osoi_ice           =>    col_ws%h2osoi_ice           , & ! Input:  [real(r8) (:,:) ]  ice lens (kg/m2)
+          h2osoi_liq           =>    col_ws%h2osoi_liq           , & ! Input:  [real(r8) (:,:) ]  liquid water (kg/m2)
+          h2osno               =>    col_ws%h2osno               , & ! Input:  [real(r8) (:)   ]  snow water (mm H2O)
+          snow_depth           =>    col_ws%snow_depth           , & ! Input:  [real(r8) (:)   ]  snow height (m)
+          h2osfc               =>    col_ws%h2osfc               , & ! Output: [real(r8) (:)   ]  surface water (mm)
+          h2orof               =>    col_ws%h2orof               , & ! Output:  [real(r8) (:)   ]  floodplain inudntion volume at column level (mm)
```
[the remark this answers is not in the document]
Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyMod.F90:337]: Yes, I will remove "at column level".

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 9]
```diff
@@ -492,7 +523,36 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
              !7. remove drainage from h2osfc and add to qflx_infl
              h2osfc(c) = h2osfc(c) - qflx_h2osfc_drain(c) * dtime
              qflx_infl(c) = qflx_infl(c) + qflx_h2osfc_drain(c)
-             qflx_gross_infl_soil(c) = qflx_gross_infl_soil(c) + qflx_h2osfc_drain(c)
+             
+             !8. add drainage from river inundation to qflx_infl (land river two way coupling)
+             if (use_lnd_rof_two_way) then
+
+               ! estimate the available space in the first soil layer for floodplain infiltration
```
[the remark this answers is not in the document]
Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyMod.F90:530]: Sounds good, I changed "space" to "volume" and added the unit [mm H2O].

File: components/mosart/bld/build-namelist [hunk 10]
```diff
@@ -312,6 +312,7 @@ else {
     add_default($nl, 'smat_option');
     add_default($nl, 'wrmflag');
     add_default($nl, 'inundflag');
+    add_default($nl, 'use_lnd_rof_two_way');
```
Comment (tanzeli1982) [inline components/mosart/bld/build-namelist:315]: Does this line and the added line in components/mosart/bld/namelist_files/namelist_defaults_mosart.xml introduce your test namelist error?

**Comment by tanzeli1982** [comment]
> This new feature is turned off by default in our development.
> 
> Land-river two-way coupling represents a realistic physical process at the river land interface (floodplain) proposed by ICoM to understand the impacts of inundation on water and biogeochemical cycles.
> 
> We documented details of the development and evaluation of this new feature in a manuscript, which is under review at JAMES. The preprint of the manuscript can be found at: [https://www.essoar.org/doi/abs/10.1002/essoar.10507802.1].
> 
> We also run the e3sm_land_developer test suites, the new development passed all the test except reporting namelist error. The namelist error is because the `use_lnd_rof_two_way=.false.` appears in **mosart_in**.

File: components/elm/cime_config/testdefs/testmods_dirs/elm/lnd_rof_2way/user_nl_elm [hunk 11]
```diff
@@ -0,0 +1 @@
+ use_lnd_rof_two_way = .true.
```
[the remark this answers is not in the document]
Comment (donghuix) [reply components/elm/cime_config/testdefs/testmods_dirs/elm/lnd_rof_2way/user_nl_elm:1]: @bishtgautam and I have discussed about how to make sure those flags from rof and lnd to be consistent, but we didn't come up with a good solution. The land river two-way coupling will not be implemented correctly if one flag is turned on but the other remains off. 

Do you know how to make sure one flag from rof is consistent with another flag from lnd? It will be great if we can let E3SM crash and report error if they are not consistent.

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 (hunk 5 above)
Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyMod.F90:411]: Thanks for the suggestion. I prefer to use the if statement, because it will be more flexible when we introduce ocean inundation into ELM in the future. For example, we need to address the condition that` frac_h2orof + frac_h2oocn + frac_h2osfc + fsno > 1`.

File: components/elm/src/biogeophys/SoilHydrologyType.F90 (hunk 7 above)
Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyType.F90:572]: This is to allow ELM to read `fdrain` from surface data. If `fdrain` is not showing up in surface data, the default `fdrain=2.5` will be used.

Comment (tanzeli1982) [reply components/elm/src/biogeophys/SoilHydrologyType.F90:572]: Do you know why fdrain was not read from surface data in the current version?

Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyType.F90:572]: I don't know the reason. So, I only let ELM to read `fdrain` from surface data when land river two-way coupling is turned on.

File: components/mosart/bld/build-namelist (hunk 10 above)
Comment (donghuix) [reply components/mosart/bld/build-namelist:315]: Yes, you are right. After removing this line, `use_lnd_rof_two_way` doesn't show in `mosart_in`, and the model still runs. But should we still keep it here to be consistent with` wrmflag`,` inundflag`, `sediflag`, etc.

Comment (tanzeli1982) [reply components/mosart/bld/build-namelist:315]: I do not think it is needed. But let us see how @hydrotian and @liho745 think about it.

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 [hunk 12]
```diff
@@ -387,14 +401,24 @@ subroutine Infiltration(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, f
                 qflx_evap(c)=qflx_ev_soil(c)
              endif
 
+             !0. partition grid-level floodplain inundation volume and fraction to each column
+             if (use_lnd_rof_two_way) then
+                h2orof(c)      = atm2lnd_vars%h2orof_grc(g) * wtgcell(c)
+                frac_h2orof(c) = atm2lnd_vars%frac_h2orof_grc(g) * wtgcell(c)
+                ! TODO: add inundfrac from ocean 
+                if ( frac_h2orof(c) > 1 - fsno - frac_h2osfc(c) ) then
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SoilHydrologyMod.F90:409]: Use `1.0_r8` instead of `1`

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 13]
```diff
@@ -2209,6 +2215,17 @@ subroutine Rtmrun(rstwr,nlend,rdate)
              !budget_terms(bv_fp_i, 1) = budget_terms(bv_fp_i, 1) + rtmCTL%inundwf(nr)        ! 17-6-7
            endif
 
+           ! land river two way coupling, update floodplain inundation volume with drainage from lnd
+           if (use_lnd_rof_two_way) then
+             TRunoff%wf_ini(nr) = TRunoff%wf_ini(nr) - rtmCTL%inundinf(nr) * coupling_period
+
+             if ( TRunoff%wf_ini(nr) < 0 ) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:2222]: Use `0._r8`

File: components/elm/src/biogeophys/SoilHydrologyMod.F90 (hunk 12 above)
Comment (donghuix) [reply components/elm/src/biogeophys/SoilHydrologyMod.F90]: Yes.

File: components/mosart/src/riverroute/RtmMod.F90 (hunk 13 above)
Comment (donghuix) [reply components/mosart/src/riverroute/RtmMod.F90]: Yes.

**Comment by tanzeli1982** [comment]
@donghuix I do not have any more comment.

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 14]
```diff
@@ -2209,6 +2215,17 @@ subroutine Rtmrun(rstwr,nlend,rdate)
              !budget_terms(bv_fp_i, 1) = budget_terms(bv_fp_i, 1) + rtmCTL%inundwf(nr)        ! 17-6-7
            endif
 
+           ! land river two way coupling, update floodplain inundation volume with drainage from lnd
+           if (use_lnd_rof_two_way) then
+             TRunoff%wf_ini(nr) = TRunoff%wf_ini(nr) - rtmCTL%inundinf(nr) * coupling_period
```
[the remark this answers is not in the document]
Comment (hydrotian) [reply components/mosart/src/riverroute/RtmMod.F90:2220]: Thanks for the explanation.

File: components/elm/cime_config/testdefs/testmods_dirs/elm/lnd_rof_2way/user_nl_elm (hunk 11 above)
[the remark this answers is not in the document]
Comment (hydrotian) [reply components/elm/cime_config/testdefs/testmods_dirs/elm/lnd_rof_2way/user_nl_elm:1]: No I don't. It would be nice to have something like that.

**Review by hydrotian — APPROVED** [review]

**Review by peterdschwartz — APPROVED** [review]

File: components/elm/src/main/elm_varctl.F90 [hunk 15]
```diff
@@ -508,6 +508,7 @@ module elm_varctl
    ! land river two way coupling
    !----------------------------------------------------------
    logical, public :: use_lnd_rof_two_way = .false.
+   integer, public :: lnd_rof_coupling_nstep = 6
```
Comment (bishtgautam) [inline components/elm/src/main/elm_varctl.F90:511]: Let's not initialize it to `0` and have a check to ensure that a valid value is read from the `lnd_in`.

Comment (donghuix) [reply components/elm/src/main/elm_varctl.F90:511]: Do you mean initialize `lnd_rof_coupling_nstep` as 0?

Comment (bishtgautam) [reply components/elm/src/main/elm_varctl.F90:511]: Add the check for `lnd_rof_coupling_nstep` at https://github.com/E3SM-Project/E3SM/blob/2a1ef482627fbb2e697158908fbb5e58e8d8620d/components/elm/src/main/controlMod.F90#L510

Comment (donghuix) [reply components/elm/src/main/elm_varctl.F90]: Done.

**Comment by rljacob** [comment]
@liho745 please review this PR.

**Review by liho745 — APPROVED** [review]

File: components/mosart/bld/build-namelist (hunk 10 above)
Comment (rljacob) [reply components/mosart/bld/build-namelist:315]: what is this flag supposed to do?

Comment (donghuix) [reply components/mosart/bld/build-namelist]: It is used to control if land-river two-way coupling is on or off. In default, it is off. But I have tested it is not necessary to set it .false. in the default MOSART namelist. So, I have made a commit to remove it from here.

Comment (rljacob) [reply components/mosart/bld/build-namelist]: Those kind of control mechanisms usually go in the coupler but it depends on what all has to happen to turn the coupling on.

Comment (donghuix) [reply components/mosart/bld/build-namelist]: Hi @rljacob, it makes sense to control it on coupler. But in this development, both ELM and MOSART need to know if two-way coupling is turned on or not. But we don't know how address this without using namelist of ELM and MOSART. Do you want to us to address it in this PR?

Comment (rljacob) [reply components/mosart/bld/build-namelist]: You put the flag in the driver namelist and then the coupler reports its value to both models via the infobuffer.  Its very error prone to rely on two different models to set the same flag.

Comment (donghuix) [reply components/mosart/bld/build-namelist]: @rljacob ,I just made a commit to remove setting use_lnd_rof_two_way from both ELM and MOSART namelist. Instead, I put this flag into `drv_in`, and let ELM and MOSART to read if two-way coupling is turn on or not from coupler. Please let me know if that commit makes sense to you.

**Comment by rljacob** [comment]
Please edit the description of the PR at the top.  It's no longer accurate.

File: components/mosart/bld/build-namelist (hunk 10 above)
Comment (donghuix) [reply components/mosart/bld/build-namelist]: Thank you, @rljacob. We just found the test doesn't pass `COMPARE_base_rest` with the default coupling configuration, such as the ELM is coupled to the coupler at 30min and MOSART at 3hr. I guess this is coupler issue, because the test passed when I set both ELM and MOSART coupled to coupler at 30 min. We are working on debugging it, will update as soon as possible.

Comment (rljacob) [reply components/mosart/bld/build-namelist]: How long is the ERS test run for?  If you don't modify it, its a few days.  If you try to do a few timesteps, it could fail if its less then that 3 hour mosart coupling frequency.   The restart frequency has to be a multiple of the longest coupling frequency.

Comment (donghuix) [reply components/mosart/bld/build-namelist]: It is 12 days for my ERS test run. So, it is long enough for the 3 hour mosart coupling frequency.

**Comment by donghuix** [comment]
Hi @rljacob, we fixed the bug for the restart issue. And the test passes with default coupling configuration.
